### PR TITLE
[ci] [gh-actions] Run the platform job using Github Actions

### LINKED ISSUE #6807: Update dev/ci/README.md to reflect current situation
- [x] The current documentation for CI at dev/ci/README.md should be updated to reflect the current situation with CircleCI and Travis.
- [x] Reference to CI in CONTRIBUTING should probably be made platform-agnostic.
- [x] https://github.com/coq/coq/wiki/Jenkins should be merged into or at least linked from the README.
- [ ] Document macOS / Windows runners.
- [x] Update dev/ci/docker/README.md to reflect the current situation.

I'm assigning to myself not to forget about it but feel free to take over.

### LINKED ISSUE #7428: Use docker image for stable Windows packaging jobs on GitLab CI
Once #7344 gets in, GitLab CI will be used to build Windows packages (in place of AppVeyor).

There is a long-standing idea that we could make these builds more reliable by using a fixed set of dependencies that we would store somewhere. (This would also mean that the open source archive doesn't have to be regenerated at each build.)

Following #7387, I suggested trying to use Docker on Windows as well for this purpose. This would mean that GitLab CI would pull a docker image containing all the required dependencies for the build from, e.g., `dockerhub.com/coqci/coq`. (Maybe we can even get the Windows runner to cache this image.) This image would only be rebuilt once in a while (likely after each release) using a recipe that would be stored in Coq `dev/ci`.

@MSoegtropIMC has some experience with Docker, so he agreed to look into this after #7344 is merged.

### LINKED ISSUE #8046: Add timing information to Windows packaging jobs.
Our windows packaging jobs are fairly slow, however profiling them seems hard due to missing time information in the logs. Please, add such information.

### LINKED ISSUE #9401: Creation of Windows workers should be documented or automated
This issue is a reminder that we should either document the creation of workers on the Inria CloudStack infrastructure, or migrate to a platform like Azure where such creation is transparent for us.

### LINKED ISSUE #11073: CI runners from ci.inria.fr need maintenance
The VMs from ci.inria.fr that we use for CI need maintenance. This is an issue per se.

On linux machines, docker images pile up. Periodically running “docker system prune” may help. Otherwise, one can list the images by executing “docker images” and remove the ones that are no longer useful by executing “docker rmi XXXX”.

On Windows machines, build directories are not always removed (usually in case of failure). I don’t know how to manually remove these directories. Maybe Michael can help.

Ping @MSoegtropIMC

### PR DESCRIPTION
**Pull request by ejgallego**
We move the CI job creating the platform windows installer from Gitlab-driven Inria custom runners to standard Github actions one.

While we are at it, we significantly cleanup the build script, trying to avoid as much hackery as possible, and splitting it in logical stages.

Notwithstanding, the current system is still not very satisfactory, in particular:

- build time is very long, we should cache cygwin folder at least, and only rebuild if the coq-platform ref we use has changed; fortunately GH actions do support this
- verbosity of the script is out of control
- we can still cleanup way more things 

Please review the list of chosen bugs below, and speak if you think it is too wide.

closes #6807 , closes #7428 , closes #8046 , closes #8622 , closes #9401 , closes #11073  

**Kind:** infrastructure.

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated
- [x] Entry added in the changelog

Note: this PR was originally started from the nice setup of Jason Gross [1] for windows CI setup, but it was updated to actually use the platform after #13598

[1] https://github.com/mit-plv/fiat-crypto/blob/master/.github/workflows/coq-windows.yml

### CODE REVIEW

**Comment by ejgallego** [comment]
I'm missing some magic here as to make this PR run the action, ummmm....

**Comment by JasonGross** [comment]
You might need to enable actions in repository settings

**Comment by ejgallego** [comment]
> **edit**, oh, maybe I just enabled local ones, and the script imports some external stuff. Let me see.

Indeed, that was it.

File: .github/workflows/windows-std.yml
Comment (gares) [inline .github/workflows/windows-std.yml:41]: can't you break long lines with `^` ?

**Comment by gares** [comment]
How does this compare with the platform CI?

**Comment by Zimmi48** [comment]
> How does this compare with the platform CI?

I was precisely hoping that this could be completed by benefiting from your experience setting up Windows CI with GitHub Actions in the coq/platform repository.

**Comment by gares** [comment]
Yes, but is this job supposed to test?

**Comment by gares** [comment]
Oh I see, this "old" PR is now just using the same job we run on Inria's windows workers right?

**Comment by gares** [comment]
I gave a quick look, it seems you don't set
```
    PLATFORM: "https://github.com/coq/platform/archive/dev-ci.zip"
    ARCH: "64"
```

**Comment by gares** [comment]
And it seems you also need some value for `CI_PROJECT_DIR`... I wander if we can just drop the .bat file and use GH actions all the way

**Comment by Zimmi48** [comment]
> Oh I see, this "old" PR is now just using the same job we run on Inria's windows workers right?

Yes, this PR is replacing the Inria workers by some GitHub Actions workers.

**Comment by gares** [comment]
OK, so maybe the question is "what is the plan for this PR"?
I got notified of a push/rebase, the header/title is outdated and the comments not relevant anymore. In short, it is another PR.
 @ejgallego you made the recent push, do you plan working on it?

**Comment by ejgallego** [comment]
> @ejgallego you made the recent push, do you plan working on it?

Sure, thanks for your help, indeed the idea now is to move the inria workers to githu actions, testing the same.

**Comment by Alizter** [comment]
@Zimmi48 Despite coqbots notifications, this PR appears to be open?

**Comment by gares** [comment]
there was a bug I believe.

**Comment by Zimmi48** [comment]
We've hit a rate limit on GitHub's side. Next occurrences will have a limit on the number of PRs that are closed in one to avoid going over this limit again.

**Comment by ejgallego** [comment]
This is failing as the windows CI has been broken for at least a week. More stuff to consider about this action:

- verbosity: current logs are way too verbose, only errors should be output
- caching: installing all the stuff on each run seems like a waste of time and energy [also too much console verbosity has bad efficiency properties]
- staging: it would be interesting if the platform build was an action itself and our CI job could inherit staging.

**Comment by ejgallego** [comment]
Now failing with 
```
<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[cairo2.0.6.2] downloaded from https://github.com/Chris00/ocaml-cairo/releases/download/0.6.2/cairo2-0.6.2.tbz
[csexp.1.5.1] downloaded from https://github.com/ocaml-dune/csexp/releases/download/1.5.1/csexp-1.5.1.tbz
[dep-gtksourceview3.24.11] downloaded from https://download.gnome.org/sources/gtksourceview/3.24/gtksourceview-3.24.11.tar.xz
[dune.2.9.0b] downloaded from https://github.com/fdopen/dune/releases/download/2.9.0/dune.2.9.0.tar.xz
[dune-configurator.2.9.0] downloaded from https://github.com/ocaml/dune/releases/download/2.9.0/dune-2.9.0.tbz
[lablgtk3.3.1.1] downloaded from https://github.com/garrigue/lablgtk/releases/download/3.1.1/lablgtk3-3.1.1.tbz
[lablgtk3-sourceview3.3.1.1] downloaded from https://github.com/garrigue/lablgtk/releases/download/3.1.1/lablgtk3-3.1.1.tbz
[result.1.5] downloaded from https://github.com/janestreet/result/releases/download/1.5/result-1.5.tbz
[zarith.1.12] downloaded from https://github.com/ocaml/Zarith/archive/release-1.12.tar.gz
Error:  The sources of the following couldn't be obtained, aborting:
          - coq.dev
          - coqide.dev
```
...

**Comment by gares** [comment]
There is a commit up there which is green. Can't you cut there and develop the improvement in another PR?

**Comment by ejgallego** [comment]
Thanks for all the help, as this is getting ready for review / merge, I've updated the description, and cleaned up the comment section.

**Comment by ejgallego** [comment]
Something that I am not sure is if the platform build is running the test suite, @MSoegtropIMC ?

**Comment by Zimmi48** [comment]
It doesn't AFAIK and this is one thing that makes me uneasy about this change: while I'm happy to move the platform job from GitLab CI to GitHub Actions, I think we might want to keep the current GitHub Action Windows test as well (and reintroduce the test-suite, that we stopped running at some point because it was too long for Azure but would fit with the GitHub Actions setup).

**Comment by ejgallego** [comment]
> It doesn't AFAIK and this is one thing that makes me uneasy about this change: while I'm happy to move the platform job from GitLab CI to GitHub Actions, I think we might want to keep the current GitHub Action Windows test as well (and reintroduce the test-suite, that we stopped running at some point because it was too long for Azure but would fit with the GitHub Actions setup).

The platform could also run the test suite easily using `opam -t` too, so I dunno, both options are fine to me.

**Comment by MSoegtropIMC** [comment]
@ejgallego : for sure one could add an option to run the test suite. The platform CI runs a very simple "smoke test" - that is one or two selected examples from each provided package. The idea is that Coq Platform is a kind of meta build and relies on that the packages it builds are OK. The idea of the "smoke test" kit is to check if e.g. installation of packages went fine.

**Comment by ejgallego** [comment]
@MSoegtropIMC , another question, how difficult would it be to split the cygwin setup to its own script?

This way we could run it in its separate step, and have an easier time caching.

Similarly for opam, would be very interesting to separate `opam install --deps-only coq` from the actual `opam install coq`

Caching for opam should work fine and be reproducible as soon as you folks start using opam lock files.

**Comment by MSoegtropIMC** [comment]
@ejgallego : this is already supported:

- pass `-build=n` to `coq_platform_make_windows.bat` - this will just setup cygwin but do not run the script
- start up the generated cygwin and run `coq_platform_make.sh` from the copied coq_platform folder or alternatively git clone something else and start the script.

But I must admit, this is not tested much and also the cygwin setup sometimes changes.

**Comment by ejgallego** [comment]
> But I must admit, this is not tested much and also the cygwin setup sometimes changes.

Thanks for the feedback, will try!

I guess the way to make this more reliable would be to structure the platform build as a sequential call of these scripts from a top-level driver.

File: doc/changelog/11-infrastructure-and-dependencies/12425-ci+gh_actions_win.rst
Comment (Zimmi48) [inline doc/changelog/11-infrastructure-and-dependencies/12425-ci+gh_actions_win.rst:2]: That's not something that we typically document in the user-facing changelog (which is already long enough without these kinds of details). It would make more sense in the dev changelog.

File: .github/CODEOWNERS
Comment (Zimmi48) [inline .github/CODEOWNERS:26]: It could make sense to reintroduce the `@coq/windows-build-maintainers` team in the `CODEOWNERS` file for the new `dev/ci/platform` directory.

File: doc/changelog/11-infrastructure-and-dependencies/12425-ci+gh_actions_win.rst
Comment (ejgallego) [reply doc/changelog/11-infrastructure-and-dependencies/12425-ci+gh_actions_win.rst:2]: As you prefer, I was under the impression that some users were downloading the windows artifacts for master, so I signaled it.

File: .github/CODEOWNERS
Comment (ejgallego) [reply .github/CODEOWNERS:26]: Oh sure, it was an oversight by me.

File: doc/changelog/11-infrastructure-and-dependencies/12425-ci+gh_actions_win.rst
Comment (Zimmi48) [reply doc/changelog/11-infrastructure-and-dependencies/12425-ci+gh_actions_win.rst:2]: That's a good point. Maybe reformulate to put this aspect forward (and maybe do not link to the fixed issues which are not really relevant for users).

Comment (ejgallego) [reply doc/changelog/11-infrastructure-and-dependencies/12425-ci+gh_actions_win.rst:2]: Ok, I did tweak it.

File: .github/CODEOWNERS
Comment (ejgallego) [reply .github/CODEOWNERS:26]: Done, thanks.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
Will merge when CI passes.

@gares can I add you to the @coq/windows-build-maintainers team?

**Comment by gares** [comment]
no

**Comment by Zimmi48** [comment]
OK, I figured that you were the best person to be in this team (more suited than me anyway) given that you've authored most of the code that is being moved in `dev/ci/platform`. Maybe we should just remove the @coq/windows-build-maintainers team and leave this under the ownership of `@coq/ci-maintainers` then.

**Comment by gares** [comment]
I'm not fluent in .bat, and I don't even have a windows PC at hand...

**Comment by Zimmi48** [comment]
Yeah, me neither. Anyway, let's keep things as they are for the time being.

@coqbot merge now

**Pull request merged**
